src/utilities/expiration_map.rs

```diff
@@ -0,0 +1,201 @@
+// Copyright 2015 MaidSafe.net limited.
+//
+// This SAFE Network Software is licensed to you under (1) the MaidSafe.net Commercial License,
+// version 1.0 or later, or (2) The General Public License (GPL), version 3, depending on which
+// licence you accepted on initial access to the Software (the "Licences").
+//
+// By contributing code to the SAFE Network Software, or to this project generally, you agree to be
+// bound by the terms of the MaidSafe Contributor Agreement, version 1.0.  This, along with the
+// Licenses can be found in the root directory of this project at LICENSE, COPYING and CONTRIBUTOR.
+//
+// Unless required by applicable law or agreed to in writing, the SAFE Network Software distributed
+// under the GPL Licence is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.
+//
+// Please review the Licences for the specific language governing permissions and limitations
+// relating to use of the SAFE Network Software.
+
+/// A time limited map of key-value pairs.
+#[allow(unused)]
+pub struct ExpirationMap<K, V> {
+    map: ::std::collections::BTreeMap<K, (V, ::time::SteadyTime)>,
+    time_to_live: ::time::Duration,
+}
+
+#[allow(unused)] // remove when an ExpirationMap has been included elsewhere in the code
+impl<K, V> ExpirationMap<K, V> where K: PartialOrd + Ord + Clone, V: Clone {
+    /// Constructor
+    pub fn with_expiry_duration(time_to_live: ::time::Duration) -> ExpirationMap<K, V> {
+        ExpirationMap {
+            map: ::std::collections::BTreeMap::new(),
+            time_to_live: time_to_live,
+        }
+    }
+
+    /// Inserts a key-value pair into the map. Returns replaced value if key is already present.
+    pub fn insert(&mut self, key: K, value: V) -> Option<V> {
+        match self.map.insert(key, (value, ::time::SteadyTime::now())) {
+            Some((value, _)) => Some(value),
+            None => None,
+        }
+    }
+
+    /// If key exists remove it from the map and return corresponding value.
+    pub fn remove(&mut self, key: &K) -> Option<V> {
+        match self.map.remove(key) {
+            Some((value, _)) => Some(value),
+            None => None,
+        }
+    }
+
+    /// Retrieves a value for the given key if present in the map.
+    pub fn get(&mut self, key: &K) -> Option<&V> {
+        match self.map.get(key) {
+            Some(&(ref value, _)) => Some(value),
```

Comment:
should the expiration not be validated on get and ignored if expired?  similar for contains_key.  On remove we can silently remove and has_expired is then not needed.  Sounds good? :)